File: apps/ruby/test/fizz_buzz_test.rb
# frozen_string_literal: true

require_relative 'test_helper'
require_relative '../lib/fizz_buzz'
require 'stringio'

class FizzBuzzTest < Minitest::Test
  def setup
    @fizzbuzz = FizzBuzz
  end

  def test_1を渡したら文字列1を返す
    assert_equal '1', @fizzbuzz.generate(1)
  end

  def test_2を渡したら文字列2を返す
    assert_equal '2', @fizzbuzz.generate(2)
  end

  def test_3を渡したらFizzを返す
    assert_equal 'Fizz', @fizzbuzz.generate(3)
  end

  def test_5を渡したらBuzzを返す
    assert_equal 'Buzz', @fizzbuzz.generate(5)
  end

  def test_15を渡したらFizzBuzzを返す
    assert_equal 'FizzBuzz', @fizzbuzz.generate(15)
  end

  def test_1から100までの数の配列を返す
    result = @fizzbuzz.generate_list

    assert_equal 100, result.length
    assert_equal '1', result[0]
    assert_equal '2', result[1]
    assert_equal 'Fizz', result[2]
    assert_equal '4', result[3]
    assert_equal 'Buzz', result[4]
    assert_equal 'FizzBuzz', result[14]
  end

  def test_プリントする
    out = StringIO.new
    $stdout = out
    @fizzbuzz.generate_list.each { |item| puts item }
    $stdout = STDOUT

    output = out.string
    assert_includes output, '1'
    assert_includes output, 'Fizz'
    assert_includes output, 'Buzz'
    assert_includes output, 'FizzBuzz'
  end
end

File: apps/php/tests/FizzBuzzTest.php
<?php

namespace App\Tests;

use App\Domain\FizzBuzzException;
use App\FizzBuzz;
use PHPUnit\Framework\TestCase;

class FizzBuzzTest extends TestCase
{
    public function test_1を渡したら文字列1を返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('1', $fizzbuzz->generate(1));
    }

    public function test_2を渡したら文字列2を返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('2', $fizzbuzz->generate(2));
    }

    public function test_3を渡したら文字列Fizzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('Fizz', $fizzbuzz->generate(3));
    }

    public function test_5を渡したら文字列Buzzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('Buzz', $fizzbuzz->generate(5));
    }

    public function test_15を渡したら文字列FizzBuzzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('FizzBuzz', $fizzbuzz->generate(15));
    }

    public function test_6を渡したら文字列Fizzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('Fizz', $fizzbuzz->generate(6));
    }

    public function test_10を渡したら文字列Buzzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('Buzz', $fizzbuzz->generate(10));
    }

    public function test_30を渡したら文字列FizzBuzzを返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $this->assertSame('FizzBuzz', $fizzbuzz->generate(30));
    }

    public function test_1から100までのFizzBuzz配列を返す(): void
    {
        $fizzbuzz = new FizzBuzz();
        $result = $fizzbuzz->generateList();

        $this->assertSame('1', $result[0]);
        $this->assertSame('Fizz', $result[2]);
        $this->assertSame('Buzz', $result[4]);
        $this->assertSame('FizzBuzz', $result[14]);
        $this->assertSame('Buzz', $result[99]);
        $this->assertCount(100, $result);
    }

    public function test_FizzBuzzをプリントする(): void
    {
        $fizzbuzz = new FizzBuzz();

        ob_start();
        $fizzbuzz->printFizzBuzz();
        $output = ob_get_clean();

        $lines = explode("\n", trim($output));
        $this->assertSame('1', $lines[0]);
        $this->assertSame('Fizz', $lines[2]);
        $this->assertSame('Buzz', $lines[4]);
        $this->assertSame('FizzBuzz', $lines[14]);
        $this->assertSame('Buzz', $lines[99]);
        $this->assertCount(100, $lines);
    }

    public function test_ファクトリメソッドでタイプ1を生成する(): void
    {
        $type = FizzBuzz::create(1);
        $result = $type->generate(3);

        $this->assertSame('Fizz', $result->getValue());
    }

    public function test_不正なタイプで例外を発生する(): void
    {
        $this->expectException(FizzBuzzException::class);

        FizzBuzz::create(4);
    }

    public function test_不正なタイプでカスタム例外を発生する(): void
    {
        $this->expectException(FizzBuzzException::class);
        $this->expectExceptionMessage('タイプ99は見つかりません');

        FizzBuzz::create(99);
    }
}
